### PR TITLE
NAS-3058 - implementation of workspace managing

### PR DESCRIPTION
**Pull request by hkad98**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py [hunk 1]
```diff
@@ -6,24 +6,53 @@
...
+                    ),
+                    relationships=JsonApiWorkspaceInRelationships(
+                        parent=JsonApiWorkspaceInRelationshipsParent(
+                            data=JsonApiWorkspaceToOneLinkage(id=self.parent_id)
+                        )
+                    ),
+                )
             )
-        )
+        else:
```
Comment (jaceksan) [inline gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py:45]: Setup relationships first.
Then create JsonApiWorkspaceInDocument once and embed relationships like **relationships.
If relationships are empty, nothing is embedded

File: gooddata-sdk/gooddata_sdk/catalog/workspace/service.py [hunk 2]
```diff
@@ -24,6 +26,59 @@
...
+    def list_workspaces(self) -> List[CatalogWorkspace]:
+        get_workspaces = functools.partial(
+            self._entities_api.get_all_entities_workspaces,
+            include=["workspaces"],
+            _check_return_type=False,
+        )
+        workspaces = load_all_entities(get_workspaces)
+        return [CatalogWorkspace.from_api(w) for w in workspaces.data]
+
+    def create_or_update(self, workspace: CatalogWorkspace) -> None:
```
Comment (jaceksan) [inline gooddata-sdk/gooddata_sdk/catalog/workspace/service.py:43]: I do not see create scenario implemented, only update and error scenarios.

**Review by jaceksan — CHANGES_REQUESTED** [review]

File: gooddata-sdk/gooddata_sdk/catalog/workspace/service.py (hunk 2 above)
Comment (hkad98) [reply gooddata-sdk/gooddata_sdk/catalog/workspace/service.py:43]: Line 58 implements create workspace scenario (if workspace is not found -> NotFoundException is raised and create is executed).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py [hunk 3]
```diff
@@ -6,24 +6,44 @@
...
+from gooddata_metadata_client.model.json_api_workspace_in_relationships import JsonApiWorkspaceInRelationships
+from gooddata_metadata_client.model.json_api_workspace_in_relationships_parent import (
+    JsonApiWorkspaceInRelationshipsParent,
+)
+from gooddata_metadata_client.model.json_api_workspace_to_one_linkage import JsonApiWorkspaceToOneLinkage
 from gooddata_sdk.catalog.entity import CatalogNameEntity
 
 
 class CatalogWorkspace(CatalogNameEntity):
+    def __init__(self, workspace_id: str, name: str, parent_id: str = None):
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py:18]: Missing return type, i.e. `-> None`

File: gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py [hunk 4]
```diff
@@ -6,24 +6,44 @@
...
+        self.parent_id = parent_id
+
     @classmethod
     def from_api(cls, entity: dict[str, Any]) -> CatalogWorkspace:
         ea = entity["attributes"]
+        er = entity.get("relationships")
         return cls(
-            id=entity["id"],
-            name=ea["name"],
+            workspace_id=entity["id"], name=ea["name"], parent_id=er.get("parent").get("data").get("id") if er else None
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py:27]: Schema of response is designed in a way that `parent` key of `relationships` is optional. `er.get("parent").get...` could fail with `AttributeError: 'NoneType' object has no attribute 'get'`.

File: gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py [hunk 5]
```diff
@@ -6,24 +6,44 @@
...
                 attributes=JsonApiWorkspaceInAttributes(
                     name=self.name,
                 ),
+                **kwargs
             )
         )
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CatalogWorkspace):
+            return NotImplemented
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py:48]: This should be `return False`. If `other` is not instance of `CatalogWorkspace` it is not equal to it.

File: gooddata-sdk/gooddata_sdk/catalog/workspace/service.py [hunk 6]
```diff
@@ -24,6 +26,59 @@
...
+    def delete_workspace(self, workspace_id: str) -> None:
+        """
+        This method is implemented according to our implementation of delete workspace,
+        which returns HTTP 204 no matter if the workspace_id exists.
+        """
+        workspaces = self.list_workspaces()
+        if workspace_id not in [w.id for w in workspaces]:
+            raise ValueError(f"Can not delete {workspace_id} workspace. This workspace does not exist.")
+        if workspace_id in [w.parent_id for w in workspaces]:
+            children = [w.id for w in workspaces if w.parent_id == workspace_id]
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/workspace/service.py:74]: Replace with slightly more efficient:
```
children = [w.id for w in workspaces if w.parent_id == workspace_id]
if children:
    raise ValueError....
```

File: gooddata-sdk/gooddata_sdk/catalog/workspace/service.py [hunk 7]
```diff
@@ -24,6 +26,59 @@
...
 
+class CatalogWorkspaceService:
+    def __init__(self, api_client: GoodDataApiClient) -> None:
+        self._client = api_client
+        self._entities_api = metadata_apis.EntitiesApi(api_client.metadata_client)
+
+    def list_workspaces(self) -> List[CatalogWorkspace]:
+        get_workspaces = functools.partial(
+            self._entities_api.get_all_entities_workspaces,
+            include=["workspaces"],
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/workspace/service.py:37]: Is it necessary to list workspaces with `include`? I do not see any usage of it. Do I overlook something?

**Review by pcerny — CHANGES_REQUESTED** [review]

File: gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py (hunk 4 above)
Comment (hkad98) [reply gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py:27]: I discussed this with @jaceksan and when workspace has relationships field then it definitely has the following structure (parent -> data –> id).

File: gooddata-sdk/gooddata_sdk/catalog/workspace/service.py (hunk 7 above)
Comment (hkad98) [reply gooddata-sdk/gooddata_sdk/catalog/workspace/service.py:37]: When there include is not presented then the call does not return information about parent.

File: gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py (hunk 3 above)
Comment (pcerny) [reply gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py:18]: Hmmm...it looks like mypy is no longer require typing `__init__` return type, see https://mypy.readthedocs.io/en/stable/class_basics.html#annotating-init-methods.
Taking my comment back even if it is more clear to type it, it is not required so lets leave it as it is.

File: gooddata-sdk/gooddata_sdk/catalog/workspace/service.py (hunk 7 above)
Comment (pcerny) [reply gooddata-sdk/gooddata_sdk/catalog/workspace/service.py:37]: Thanks for clarification.

File: gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py (hunk 4 above)
Comment (pcerny) [reply gooddata-sdk/gooddata_sdk/catalog/workspace/model/workspace.py:27]: Schema is not in our favor. It seems we are 100% sure it is implemented in GD.CN the way you mentioned. Consider this comment covered.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Review by pcerny — APPROVED** [review]

**Review by jaceksan — APPROVED** [review]

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-sonarqube-scan-zuul-gate/67/ : SUCCESS in 1m 04s
- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-fossa-scanning-zuul-el7/67/ : FAILURE in 2m 18s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-merge-compliance-check/67/ : SUCCESS in 19s

**Pull request merged**